File: 탐욕법/체육복/solve.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    int answer = n - lost.size();
    int cnt = 0;
    
    sort(lost.begin(), lost.end());
    sort(reserve.begin(), reserve.end());
      
    vector<int> reserved;
    for(int i=0; i<reserve.size(); i++){
        bool flag = true;
        
        for(int j=0; j<lost.size(); j++){
            if(reserve[i] == lost[j]){
                lost.erase(lost.begin()+j);
                cnt++;
                flag = false;
                break;
            }
        }
        
        if(flag) reserved.push_back(reserve[i]);
    }

    for(int i=0; i<reserved.size(); i++){
        int cur = reserved[i];
        
        for(int j=0; j<lost.size(); j++){
            if(lost[j] == cur-1){
                cnt++;
                lost.erase(lost.begin()+j);
                break;
            } else if(lost[j] == cur+1){
                cnt++;
                lost.erase(lost.begin()+j);
                break;
            }
        }   
    }
    answer += cnt;
    
    return answer;
}

File: 이분탐색/징검다리/solve.py
def bisect(rocks, n, l, r):
    if l+1 >= r:
        return l
    
    mid = (int)((l+r)/2)
    
    before = 0
    cnt = 0
    for rock in rocks:
        if rock-before < mid:
            cnt += 1
        else :
            before = rock
    
    if cnt <= n:
        return bisect(rocks, n, mid, r)
    else :
        return bisect(rocks, n, l, mid)
        

def solution(distance, rocks, n):
    rocks.append(distance)
    rocks.sort()
    
    answer = bisect(rocks, n, 0, distance)
    return answer
